File: Metro/src/main/java/org/cubexmc/metro/bedrock/BedrockDetector.java
package org.cubexmc.metro.bedrock;

import java.lang.reflect.Method;
import java.util.UUID;

import org.bukkit.entity.Player;

/**
 * Optional Bedrock-player detection through Geyser/Floodgate without a hard API dependency.
 */
final class BedrockDetector {

    private BedrockDetector() {
    }

    static boolean isBedrockPlayer(Player player) {
        if (player == null) {
            return false;
        }
        UUID uuid = player.getUniqueId();
        return isGeyserPlayer(uuid) || isFloodgatePlayer(uuid);
    }

    private static boolean isGeyserPlayer(UUID uuid) {
        try {
            Class<?> apiClass = Class.forName("org.geysermc.geyser.api.GeyserApi");
            Method apiMethod = apiClass.getMethod("api");
            Object api = apiMethod.invoke(null);
            if (api == null) {
                return false;
            }
            Method isBedrockPlayer = apiClass.getMethod("isBedrockPlayer", UUID.class);
            return Boolean.TRUE.equals(isBedrockPlayer.invoke(api, uuid));
        } catch (ReflectiveOperationException | LinkageError ignored) {
            return false;
        }
    }

    private static boolean isFloodgatePlayer(UUID uuid) {
        try {
            Class<?> apiClass = Class.forName("org.geysermc.floodgate.api.FloodgateApi");
            Method getInstance = apiClass.getMethod("getInstance");
            Object api = getInstance.invoke(null);
            if (api == null) {
                return false;
            }
            Method isFloodgatePlayer = apiClass.getMethod("isFloodgatePlayer", UUID.class);
            return Boolean.TRUE.equals(isFloodgatePlayer.invoke(api, uuid));
        } catch (ReflectiveOperationException | LinkageError ignored) {
            return false;
        }
    }
}

File: Metro/src/main/java/org/cubexmc/metro/bedrock/BedrockDetector.kt
package org.cubexmc.metro.bedrock

import java.util.UUID
import org.bukkit.entity.Player

/**
 * Optional Bedrock-player detection through Geyser/Floodgate without a hard API dependency.
 */
object BedrockDetector {
    @JvmStatic
    fun isBedrockPlayer(player: Player?): Boolean {
        if (player == null) {
            return false
        }
        val uuid = player.getUniqueId() ?: return false
        return isGeyserPlayer(uuid) || isFloodgatePlayer(uuid)
    }

    private fun isGeyserPlayer(uuid: UUID): Boolean {
        try {
            val apiClass = Class.forName("org.geysermc.geyser.api.GeyserApi")
            val apiMethod = apiClass.getMethod("api")
            val api = apiMethod.invoke(null) ?: return false
            val isBedrockPlayer = apiClass.getMethod("isBedrockPlayer", UUID::class.java)
            return java.lang.Boolean.TRUE == isBedrockPlayer.invoke(api, uuid)
        } catch (ignored: ReflectiveOperationException) {
            return false
        } catch (ignored: LinkageError) {
            return false
        }
    }

    private fun isFloodgatePlayer(uuid: UUID): Boolean {
        try {
            val apiClass = Class.forName("org.geysermc.floodgate.api.FloodgateApi")
            val getInstance = apiClass.getMethod("getInstance")
            val api = getInstance.invoke(null) ?: return false
            val isFloodgatePlayer = apiClass.getMethod("isFloodgatePlayer", UUID::class.java)
            return java.lang.Boolean.TRUE == isFloodgatePlayer.invoke(api, uuid)
        } catch (ignored: ReflectiveOperationException) {
            return false
        } catch (ignored: LinkageError) {
            return false
        }
    }
}
